src/java_bytecode/java_bytecode_convert_class.cpp

```diff
@@ -86,55 +86,67 @@ class java_bytecode_convert_classt:public messaget
 };
 
 /// Auxiliary function to extract the generic superclass reference from the
-/// class signature. If the superclass is not generic, it returns an empty
-/// string.
+/// class signature. If the signature is empty or the superclass is not generic
+/// it returns empty.
+/// Example:
+/// - class: A<T> extends B<T, Integer> implements C, D<T>
+/// - signature: <T:Ljava/lang/Object;>B<TT;Ljava/lang/Integer;>;LC;LD<TT;>;
+/// - superclass reference: B<TT;Ljava/lang/Integer;>;
 /// \param signature Signature of the class
 /// \return Reference of the generic superclass, or empty if the superclass
 /// is not generic
-static std::string
-extract_generic_superclass_reference(const std::string &signature)
+static optionalt<std::string>
+extract_generic_superclass_reference(const optionalt<std::string> &signature)
 {
-  // skip the (potential) list of generic parameters at the beginning of the
-  // signature
-  const size_t start = signature.front() == '<'
-                         ? find_closing_delimiter(signature, 0, '<', '>') + 1
-                         : 0;
-
-  // extract the superclass reference
-  const std::string superclass_ref = signature.substr(
-    start,
-    find_closing_semi_colon_for_reference_type(signature, start) - start + 1);
-
-  // if the superclass is generic then the reference is of form
-  // Lsuperclass-name<generic-types>;
-  if(superclass_ref.substr(superclass_ref.length() - 2) == ">;")
-    return superclass_ref;
-  else
-    return "";
+  if(signature.has_value())
+  {
+    // skip the (potential) list of generic parameters at the beginning of the
+    // signature
+    const size_t start =
+      signature.value().front() == '<'
+        ? find_closing_delimiter(signature.value(), 0, '<', '>') + 1
+        : 0;
+
+    // extract the superclass reference
+    const std::string superclass_ref = signature.value().substr(
+      start,
+      find_closing_semi_colon_for_reference_type(signature.value(), start) -
```

Comment:
Nit: the minus trailing at the end is a little hard to read, I'd prefer pulling `find_closing_semi_colon_for_reference_type(signature.value(), start) ` into a variable if needed to get param on one line